### PR TITLE
Decouple the OpenID module from YesSql and introduce OrchardCore.OpenId.EntityFrameworkCore

### PR DESCRIPTION
**Pull request by PinpointTownes**
Similarly to what's been done for the OrchardCore.Users module, this PR decouples the OpenID module (and more specifically its 2 controllers) from the YesSql entities so the YesSql stores can be replaced by different implementations (this PR includes a new `OrchardCore.OpenId.EntityFrameworkCore` based on the official OpenIddict EF Core 2.0 stores).

Major things that have changed:

  - New (empty) models abstractions/markers have been introduced: `IOpenIdApplication`/`IOpenIdAuthorization`/`IOpenIdScope`/`IOpenIdToken`. They are part of the main `OrchardCore.OpenId` module as introducing an abstractions package didn't seem necessary nor useful.

  - New stores abstractions have been added: `IOpenIdApplicationStore`/`IOpenIdAuthorizationStore`/`IOpenIdScopeStore`/`IOpenIdTokenStore`. They are all non-generic and derive from the generic OpenIddict stores (e.g `IOpenIdApplicationStore` derives from `IOpenIddictApplicationStore<IOpenIdApplication>`. Dealing with these abstractions is only necessary if you plan to write custom stores (i.e not based on YesSql nor EF Core).

  - New non-generic managers have been created: `OpenIdApplicationManager`/`OpenIdAuthorizationManager`/`OpenIdScopeManager`/`OpenIdTokenManager`. This allows getting rid of the generics in the controllers and exposing Orchard-specific methods that are used in the controllers.

  - The controllers have been revamped not to use the YesSql-specific entities. Instead of directly operating on the models, they rely on the application manager to update the model properties (that will itself delegate this task to the registered store).

  - New (partially) generic EF Core models and stores have been introduced in a separate `OrchardCore.OpenId.EntityFrameworkCore` module. The stores inherit from the OpenIddict EF Core stores AND **explicitly** implement the Orchard stores interfaces to perform the necessary casts. This pattern proved to be useful as one can create a custom store based on the EF Core one without being forced to deal with casts/downcasts. I plan to generalize this pattern and adopt it for both the YesSql stores and the Orchard-specific methods in the EF Core stores.

Here's an example of a custom store based on `OrchardCore.OpenId.EntityFrameworkCore`:

```csharp
public class MyStore : OpenIdApplicationStore<MyContext, string>
{
    public MyStore(MyContext context)
        : base(context)
    {
    }

    public override Task<OpenIdApplication<string>> CreateAsync(
        OpenIdApplication<string> application, CancellationToken cancellationToken)
    {
        // No cast is required as you directly operate on the concrete entity, not the abstraction.
        var name = application.DisplayName;

        return base.CreateAsync(application, cancellationToken);
    }
}
```

@sebastienros @jersiovic please let me know what you think when you have a sec' :sweat_smile: 

Note: it's still a WIP, don't merge this PR yet.

### CODE REVIEW

File: src/OrchardCore.Cms.Web/Startup.cs [hunk 1]
```diff
@@ -1,12 +1,42 @@
...
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using OrchardCore.Logging;
+using OrchardCore.OpenId.EntityFrameworkCore.Models;
 
 namespace OrchardCore.Cms.Web
 {
+
+    public class MyContext : DbContext
```
Comment (PinpointTownes) [inline src/OrchardCore.Cms.Web/Startup.cs]: Note: these changes will be reverted once this PR is ready. They are just here to show you how one can use the OpenID module as a management UI for an existing OpenIddict deployment (you basically just have to ignore the Orchard-specific model properties).

**Comment by sebastienros** [comment]
Have you tried some multi-tenancy with it? Table prefix I assume is not an option with EF, but at least different databases.

**Comment by PinpointTownes** [comment]
@sebastienros nope, not yet. It's indeed one of the things we'll have to discuss (e.g do we want to have a tenant-aware `OrchardDbContext` in a dedicate Orchard package?)

**Comment by PinpointTownes** [comment]
After our discussion on Gitter, I started prototyping a new `DbContextOptionsBuilder` extension that configures the database provider and adds the configured table prefix to the entities. I opened https://github.com/OrchardCMS/OrchardCore/issues/1343 to track that work.

**Comment by PinpointTownes** [comment]
This PR is ready to be reviewed (and merged).

File: src/OrchardCore.Modules/OrchardCore.OpenId.EntityFrameworkCore/Startup.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            }
+            catch (Exception exception)
+            {
+                _logger.LogWarning(exception, "An error occurred while registering the OpenID Entity Framework Core stores.");
+            }
+        }
+
+        private (Type applicationType, Type authorizationType, Type scopeType, Type tokenType) GetEntityTypes(Type keyType)
+        {
+            Type GetEntityType(string node)
```
[the remark this answers is not in the document]
Comment (PinpointTownes) [reply src/OrchardCore.Modules/OrchardCore.OpenId.EntityFrameworkCore/Startup.cs]: :+1:

**Comment by PinpointTownes** [comment]
> What is the purpose of adding EntityFramework stores to Orchard and implement this #1343 ? It looks it will ease the use of Orchard Framework in scenarios where we need to combine Orchard with another app based on EF sharing its db.

Yep, that's the idea. As part of the next PR, we'll also split the OpenID module into 3 separate features: management, server and token validation (e.g you'll be able to use the management feature of the Orchard OpenID module with an existing OpenIddict deployment using EF Core):

image

**Comment by sebastienros** [comment]
I think these features should mention Open Id in their names. They are too generic, and could be existing in any other component. I know there is a category, but looking at this screenshot it's far from obvious.

**Comment by PinpointTownes** [comment]
@sebastienros another PR, another discussion :sweat_smile:

**Pull request merged**